modules/framework.lua

```diff
@@ -0,0 +1,1321 @@
+end
+
+function Plugin:report(metrics)
+  self:emit('report')
+  self:onReport(metrics)
+end
+
+function Plugin:onReport(metrics)
+  -- metrics can be { metric = value }
+  -- or {metric = {value, source}}
+  -- or {metric = {{value, source}, {value, source}, {value, source}}
+  -- or {metric, value, source}
+  -- or {{metric, value, source, timestamp}}
+  for metric, v in pairs(metrics) do
+    -- { { metric, value .. }, { metric, value .. } }
+    if type(metric) == 'number' then
+      print(self:format(v.metric, v.value, v.source, v.timestamp or currentTimestamp()))
+    elseif type(v) ~= 'table' then
+      print(self:format(metric, v, self.source, currentTimestamp()))
+    elseif type(v[1]) ~= 'table' and v.value then
+      -- looking for { metric = { value, source, timestamp }}
+      local source = v.source or self.source
+      local value = v.value
+      local timestamp = v.timestamp or currentTimestamp()
+      print(self:format(metric, value, source, timestamp))
+    else
+      -- looking for { metric = {{ value, source, timestamp }}}
+      for _, j in pairs(v) do
+        local source = j.source or self.source
+        local value = j.value
+        local timestamp = j.timestamp or currentTimestamp()
+        print(self:format(metric, value, source, timestamp))
+      end
+    end
+  end
+end
+
+function Plugin:format(metric, value, source, timestamp)
+  self:emit('format')
+  return self:onFormat(metric, value, source, timestamp)
+end
+
+--- Called by the framework before formating the metric output.
+-- @string metric the metric name
+-- @param value the value to format
+-- @string source the source to report for the metric
+-- @param timestamp the time the metric was retrieved
+-- You can override this on your plugin instance.
+function Plugin:onFormat(metric, value, source, timestamp)
+  return string.format('%s %f %s %s', metric, value, source, timestamp)
+end
+
+local CommandPlugin = Plugin:extend()
+framework.CommandPlugin = CommandPlugin
+
+function CommandPlugin:initialize(params)
+  Plugin.initialize(self, params)
+
+  if not params.command then
+    error('params.command undefined. You need to define the command to excetue.')
+  end
+  self.command = params.command
+end
+
+function CommandPlugin:execCommand(callback)
+  local proc = io.popen(self.command, 'r')
+  local output = proc:read("*a")
+  proc:close()
+  if callback then
+    callback(output)
+  end
+end
+
+function CommandPlugin:onPoll()
+  self:execCommand(function (output) self.parseCommandOutput(self, output) end)
+end
+
+function CommandPlugin:parseCommandOutput(output)
+  local metrics = self:onParseCommandOutput(output)
+  self:report(metrics)
+end
+
+function CommandPlugin:onParseCommandOutput(output)
+  print(output)
+  return {}
+end
+
+local NetPlugin = Plugin:extend()
+framework.NetPlugin = NetPlugin
+
+local HttpPlugin = Plugin:extend()
+framework.HttpPlugin = HttpPlugin
+
+function HttpPlugin:initialize(params)
+  Plugin.initialize(self, params)
```

Comment:
@codemoran I'm pretty sure that in the first version of the LUA plugin that option never worked because I looked for that option inside the luvit source code but I didn't found anything about it. (I saw that option in the original .js plugin)
I can do some tests to see how we can handle this scenario using the Luvit https implementation.